### PR TITLE
Allow assignments to a narrowable reference to be considered narrowable

### LINKED ISSUE #35484: Generator<T, TReturn>['next'] doesn't narrow properly when using assignment expressions or the comma operator
**TypeScript Version:**  3.8.0-dev.20191203

**Search Terms:**

* generator
* generators
* iterator
* iterators
* narrow
* narrowing
* assignment
* expression
* next
* comma
* operator

**Code**

```ts
function* generator () {
	yield 4;
	return 'hello world';
}

function consumer1 () {
	const gen = generator();
	let result = gen.next();
	while (!result.done) {
		((x: number) => {})(result.value);
		result = gen.next();
	}
	((x: string) => {})(result.value);
}

function consumer2 () {
	const gen = generator();
	let result;
	while (!(result = gen.next()).done) {
		((x: number) => {})(result.value);
	}
	((x: string) => {})(result.value);
}

function consumer3 () {
	const gen = generator();
	let result;
	while (result = gen.next(), !result.done) {
		((x: number) => {})(result.value);
	}
	((x: string) => {})(result.value);
}
```

**Expected behavior:**

Type narrowing for `result.value` should be the same in `consumer1`, `consumer2`, and `consumer3`.

**Actual behavior:**

`consumer1` narrows `result.value`.

`consumer2` and `consumer3` don't narrow `result.value`, thus giving the following errors:

> Argument of type 'string | number' is not assignable to parameter of type 'number'.
> > Type 'string' is not assignable to type 'number' .ts(2345)

> Argument of type 'string | number' is not assignable to parameter of type 'string'.
> > Type 'number' is not assignable to type 'string'. ts(2345)

Explicitly declaring the type of `result` to be `IteratorResult<number, string>` (the type inferred in `consumer1`) does not fix the issue.

**Playground Link:**

[Playground Link](https://www.typescriptlang.org/play/?ts=3.8.0-dev.20191128#code/GYVwdgxgLglg9mAVAAgOYFMzoE4EMpzbIAUAlMgN4BQAkAJ4zoA2AJsgCwDct26UI2MMgDkAC2ZM4yAO6FWw7gF8qVUJFgJkEBAGcQAWxwBGEuWo1tYHVDSZkAXltY8BbGW40mfZLz1MbjhhgAHRYAB5Q7rTSojBeJACEviD+wSwI6Ga0NMTEYQBcyGAGAEY45PYAfJSKpMTJqQBuuEwgmR40DQFOoegRUTTKOXmF1tgwYKgV1RS19eh+UMHNre1Uyqrg0PBClnqG2ABMppS0ezZBDk44+IQDXjZdHjFx6Inzi1dBvf2kpGkZLI0YYFIqlcoOGZzLrLFptUgeIa5UFjCZTSE1OowlbwpQqNTbTR7Aw4ADMJ3M5ycX0wN1c928T2isXiHxS3W+4UipAANMgkgt2QCsECQYVivoythppi2U04WtBrRkaMoONJjLZljBfLVgj1kA)

**Related Issues:**

None

### PR DESCRIPTION
**Pull request by rbuckton**
This modifies control flow to consider an assignment to be narrowable if the assignment target is a narrowable reference.

Given:
```ts
type D = { done: true, value: 1 } | { done: false, value: 2 };
declare function f(): D;
// (A)
{
  let o: D;
  if (o = f(), o.done) {
    o.value; // '1'
  }
}
 
// (B)
{
  let o: D;
  if ((o = f()).done) {
      o.value; // '1 | 2'
  }
}
```

Prior to this fix, `o.value` in `A` had the type `1` but `o.value` in `B` had the type `1 | 2` because we did not consider `(o = f())` to be narrowable.

This also fixes some discrepancies in the control flow graph formatter used for debugging to handle circular references in loops.

Fixes #35484

### CODE REVIEW

**Comment by rbuckton** [comment]
@weswigham @ahejlsberg 
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
@typescript-bot user test this

**Comment by rbuckton** [comment]
@typescript-bot user test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20131,6 +20131,11 @@ namespace ts {
                 case SyntaxKind.ParenthesizedExpression:
                 case SyntaxKind.NonNullExpression:
                     return isMatchingReference(source, (target as NonNullExpression | ParenthesizedExpression).expression);
+                case SyntaxKind.BinaryExpression:
+                    if (isAssignmentExpression(target)) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:20135]: Couldn't this just be

```ts
return isAssignmentExpression(target) && isMatchingReference(source, target.left);
```

Comment (rbuckton) [reply src/compiler/checker.ts:20135]: Not all binary expressions that are narrowing expressions are assignment expressions. Being that exclusive might break cases using `instanceof`, `in`, or `,`.

**Comment by rbuckton** [comment]
The changes in dbg.ts were to fix a circular reference issue that I found while I was investigating the control flow graphs of the original issue and aren't directly related to this fix.

File: src/compiler/checker.ts (hunk 1 above)
Comment (rbuckton) [reply src/compiler/checker.ts:20135]: After reviewing `isMatchingReference` further, it doesn't seem like those other binary expressions can be narrowable references, so I'll go ahead with this change.

**Comment by rbuckton** [comment]
@ahejlsberg any other feedback? I'd like to get this in for 4.0.1 if its an acceptable change as this issue has already been rescheduled several times.

**Pull request merged**